### PR TITLE
`NonInteractingDeformableMolecules` disorder model and `Crystal` tools

### PR DESCRIPTION
**Pull request by fredericpoitevin**
The new `NonInteractingDeformableMolecules` (NIDM) model was added. It depends on the `GaussianNetworkModel` class newly added too, which in turn depends on the newly added `Crystal` object (to build the neighbor unit cells around the reference unit cell, mainly). Helper visualization tools were created for the latter.

At the moment, only the "soft-coupling limit" of the NIDM model has been implemented. 

The intended usage is as follows:
```python
from eryx.pdb import AtomicModel, Crystal
from eryx.visuals import VisualizeCrystal
from eryx.models import NonInteractingDeformableMolecules as NIDM

pdb_path='../tests/pdbs/6o2h_clean.pdb'

model = AtomicModel(pdb_path, expand_p1=True)

crystal = Crystal(model)
crystal.supercell_extent(nx=1, ny=1, nz=1)

visualize_crystal = VisualizeCrystal(crystal)
visualize_crystal.color_by = 'unit_cell'
visualize_crystal.show()
```

Resulting in an interactive `plotly` figure:
image

```python
hsampling, ksampling, lsampling = ((-13,13,5),(-16,16,5),(-17,17,5))
nidm = NIDM(pdb_path, hsampling, ksampling, lsampling)
Id_nidm = nidm.apply_disorder()
visualize_central_slices(np.real(Id_nidm).reshape(nidm.map_shape), vmax_scale=11
```
output

To explore a little bit under the hood, one can interactively plot the covariance matrix:
```python
import plotly.express as px
import plotly.io as pio
pio.renderers.default = 'browser'
fig = px.imshow(nidm.covar)
fig.show()
```
image

or just the modelled Atomic Displacement Parameters:
```python
fig = px.scatter(nidm.ADP)
fig.show()
```
Screen Shot 2023-01-21 at 6 10 12 PM

*Note: unit tests to follow hopefully soon*

### CODE REVIEW

**Comment by fredericpoitevin** [comment]
I withdraw the PR momentarily to add the GNM model to it.

**Comment by fredericpoitevin** [comment]
I edited the PR title and first comment to reflect the updated content.

File: eryx/pdb.py [hunk 1]
```diff
@@ -316,3 +341,308 @@ def _tile_form_factors(self):
...
+        -------
+        hessian: numpy.ndarray,
+                 shape (n_asu, n_atoms_per_asu,
+                        n_cell, n_asu, n_atoms_per_asu)
+                 type 'complex'
+            - dimension 0: index ASUs in reference cell
+            - dimension 1: index their atoms
+            - dimension 2: index neighbor cells
+            - dimension 3: index ASUs in neighbor cell
+            - dimension 4: index atoms in neighbor ASU
```
Comment (apeck12) [inline eryx/pdb.py:562]: Yikes, I didn't realize the dimensionality of the Hessian was so frighteningly complex! Sounds like a broadcasting nightmare waiting to happen...

File: eryx/pdb.py [hunk 2]
```diff
@@ -316,3 +341,308 @@ def _tile_form_factors(self):
...
+    def compute_Kinv(self, hessian, kvec=None):
+        """
+        Compute the inverse of K(kvec)
+        (see compute_K() for the relationship between K and the hessian).
+
+        Parameters
+        ----------
+        hessian : numpy.ndarray, see compute_hessian()
+        kvec : numpy.ndarray, shape (3,)
+            phonon wavevector, default array([0.,0.,0.])
```
Comment (apeck12) [inline eryx/pdb.py:632]: For my understanding: a kvec of (0,0,0) implies no lattice dynamics? (And not that this is being evaluated at an integral Miller index?)

File: eryx/scatter.py [hunk 3]
```diff
@@ -63,11 +66,15 @@ def structure_factors_batch(q_grid, xyz, ff_a, ff_b, ff_c, U=None, sum_over_atom
     
     A = 1j * fj * np.sin(np.dot(q_grid, xyz.T)) * np.exp(-0.5 * qUq)
     A += fj * np.cos(np.dot(q_grid, xyz.T)) * np.exp(-0.5 * qUq)
+    if project_on_components is not None:
```
Comment (apeck12) [inline eryx/scatter.py:69]: Neat!

File: eryx/visuals.py [hunk 4]
```diff
@@ -32,3 +34,134 @@ def visualize_central_slices(I, vmax_scale=5):
...
+        """
+        Plotly helper functions to visualize the Crystal object.
+        Parameters
+        ----------
+        crystal : eryx.models.Crystal object
+        """
+        self.crystal = crystal
+        self.draw_data = None
+        self.color_by = 'asu_id'
+        self.color_palette = 'xkcd'
```
Comment (apeck12) [inline eryx/visuals.py:51]: Haha, so great!

**Review by apeck12 — APPROVED** [review]
Thanks for all these new features, Fred!!

File: eryx/pdb.py (hunk 1 above)
Comment (fredericpoitevin) [reply eryx/pdb.py:562]: Yes I've been back and forth with the actual order and ended up with this one - hopefully it won't bite me later!

File: eryx/pdb.py (hunk 2 above)
Comment (fredericpoitevin) [reply eryx/pdb.py:632]: Excellent question, let me see if I understand it enough to detail things a bit below:

Here, `kvec=(0,0,0)` is an arbitrary choice for a default vector. 

In the case where we consider interactions across asymmetric units, it does correspond to the wavevector of acoustic phonons and more particularly to translations of the crystal as a whole (longest wavelength), which indeed does not change interatomic distances and only contribute at integral Miller indices.

In the NIDM model, pairwise interactions across unit cells are zero in the first place. This is implemented by setting the corresponding entries in the hessian matrix $\mathbf{H}$ to zero. Then, the matrix 
```math
\mathbf{K}(\mathbf{k}) = \sum_{c} \mathbf{H}_{0,c} e^{i\mathbf{k}^T \mathbf{r}}
```
reduces to just the Hessian of the reference unit cell $\mathbf{H_{0,0}}$ and thus independent of $\mathbf{k}$. So in the NIDM model we could actually compute the generalized inverse of $\mathbf{K}$ at any wavevector $\mathbf{k}$ and get the same result: the covariance matrix - which will be zero for every pair of atoms that do not belong to the same unit cell.

Comment (apeck12) [reply eryx/pdb.py:632]: Got it. Thanks for that explanation!

**Pull request merged**